jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,441 @@
+    }
+
+    public final void setMaxMultiplex(int maxMultiplex)
+    {
+        if (maxMultiplex < 1)
+            throw new IllegalArgumentException("Max multiplex must be >= 1");
+        this.maxMultiplex = maxMultiplex;
+    }
+
+    public int getMaxUsageCount()
+    {
+        return maxUsageCount;
+    }
+
+    public final void setMaxUsageCount(int maxUsageCount)
+    {
+        if (maxUsageCount == 0)
+            throw new IllegalArgumentException("Max usage count must be != 0");
+        this.maxUsageCount = maxUsageCount;
+    }
+
+    /**
+     * Create a new disabled slot into the pool. The returned entry
+     * won't be acquirable as long as {@link Entry#enable(Object)}
+     * has not been called.
+     *
+     * @param maxPending the max desired number of connections scheduled for opening,
+     * or a negative number to always trigger the opening of a new connection
+     * @return a disabled entry that is contained in the pool,
+     * or null if the pool is closed.
+     */
+    public Entry addReservation(int maxPending)
+    {
+        closeLock.lock();
+        try
+        {
+            long pending = sharedList.stream().filter(entry -> entry.getPooled() == null).count();
+            if (maxPending >= 0 && pending >= maxPending)
+            {
+                return null;
+            }
+
+            if (!closed && sharedList.size() < maxEntries)
+            {
+                Entry entry = new Entry();
+                sharedList.add(entry);
+                return entry;
+            }
+            return null;
+        }
+        finally
+        {
+            closeLock.unlock();
+        }
+    }
+
+    /**
+     * Acquire the entry from the pool at the specified index.
+     *
+     * @param idx the index of the entry to acquire.
+     * @return the specified entry or null if there is none at the specified index or if it is not available.
+     */
+    public Entry acquireAt(int idx)
+    {
+        if (closed)
+            return null;
+
+        try
+        {
+            Entry entry = sharedList.get(idx);
+            if (entry.tryAcquire())
+                return entry;
+        }
+        catch (IndexOutOfBoundsException e)
+        {
+            // no entry at that index
+        }
+        return null;
+    }
+
+    /**
+     * Acquire an entry from the pool.
+     *
+     * @return an entry from the pool or null if none is available.
+     */
+    public Entry acquire()
+    {
+        if (closed)
+            return null;
+
+        // first check the thread-local cache
+        if (cache != null)
+        {
+            List<Entry> cachedList = cache.get();
+            while (!cachedList.isEmpty())
+            {
+                Entry cachedEntry = cachedList.remove(cachedList.size() - 1);
+                if (cachedEntry.tryAcquire())
+                    return cachedEntry;
+            }
+        }
+
+        // then iterate the shared list
+        for (Entry entry : sharedList)
```

Comment:
Well, this behavior was chosen on purpose: always try to reuse the same connections to let the lesser used connections cool down and eventually timeout and shrink the pool. The sibling `acquireAt(int)` method is there in case you want to implement exactly what you describe externally, which is what the `RoundRobinConnectionPool` does.